source/arm9/math.c

```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: Zlib
+//
+// Copyright (C) 2024 Dominik Kurz
+#include <nds/arm9/math.h>
+
+
+float hw_sqrtf(float x)
+{
+    union{float f; u32 i;}xu;
+    xu.f = x;
+    u32 mantissa32 = xu.i & ((1 << 23) - 1); // extracts mantissa via bitmask
+    mantissa32 += (1 << 23); // adds implicit bit to mantissa.
+    // If you were planning to handle subnormal numbers 
+    // you would check if the number is subnormal before 
+    // adding the implicit bit.
+    // You would also need to do the shift of the mantissa and exponent
+    // differently, probably should use __builtin_clz for this
+    // as the count leading zeroes (clz) instruction is 1 cycle on arm
+    // but this is not my problem
+    // performance-critical code has subnormals turned off
+    // due to -ffast-math 
+    u64 mantissa = mantissa32; // cast so that the shift doesnt go wrong 
+    mantissa <<= (((xu.i & (1 << 23)) == 0) + 25);
+    // applies additional shift depending on whether exponent is even or odd
+    // 25 is 23+2, the +2 is necessary for rounding
+    REG_SQRTCNT = SQRT_64;  
+    REG_SQRT_PARAM = mantissa;
+    // starts hardware squareroot 
+    // It is critical that this happens as early as possible so that
+    // we have time to do other stuff in the meantime
+    u32 raw_exponent = (xu.i & (0xff << 23));
+    u32 sign = (xu.i & (1 << 31));
+    // check if exponent is 0
+    if (raw_exponent != 0)  
+    {
+        if (sign) //check if negative
+        {
```

Comment:
While you're at it, can you indent the comment below 4 more spaces to the right? Don't create a new commit for that, just squash all commits and leave one only commit with the title I've mentioned.